src/result.ts

```diff
@@ -351,19 +351,91 @@ export class AsyncResult<Value, Err> extends Promise<Result<Value, Err>> {
+	 * The {@link transform} callback can also return other {@link Result} or {@link AsyncResult} instances,
+	 * which will be returned as-is (the `Error` types will be merged).
+	 * The operation will be ignored if the result represents a success.
+	 *
+	 * @param transform callback function to transform the error of the result. The callback can be async as well.
+	 * @returns a new {@linkcode AsyncResult} instance with the transformed error
+	 *
+	 * > [!NOTE]
+	 * > Any exceptions or rejections that might be thrown inside the {@link transform} callback are not caught, so it is your responsibility
+	 * > to handle these exceptions. Please refer to {@linkcode AsyncResult.mapCatching} for a version that catches exceptions
+	 * > and encapsulates them in a failed result.
+	 *
+	 * @example
+	 * transforming the value of a result
+	 * ```ts
+	 * declare const result: AsyncResult<number, Error>;
+	 *
+	 * const transformed = result.mapError((error) => new CustomError()); // AsyncResult<number, CustomError>
+	 * ```
+	 *
+	 * @example
+	 * returning a result instance
+	 * ```ts
+	 * declare const result: AsyncResult<number, Error>;
+	 * declare function createCustomError(error: unknown): Result<never, CustomError>;
+	 *
+	 * const transformed = result.mapError((error) => createCustomError(error)); // AsyncResult<number, CustomError>
+	 * ```
+	 */
+	mapError<ReturnType>(transform: (error: Err) => ReturnType) {
+		return new AsyncResult<any, any>((resolve, reject) =>
+			this.then((result) => {
+				if (result.isError()) {
+					try {
+						const returnValue = transform((result as { error: Err }).error);
+						if (isPromise(returnValue)) {
+							returnValue
+								.then((value) =>
+									resolve(Result.isResult(value) ? value : Result.error(value)),
+								)
+								.catch(reject);
+						} else {
+							resolve(
+								Result.isResult(returnValue)
+									? returnValue
+									: Result.error(returnValue),
+							);
+						}
+					} catch (error) {
+						reject(error);
+					}
+				} else {
+					resolve(result);
+				}
+			}),
+		) as ReturnType extends Promise<infer PromiseValue>
+			? PromiseValue extends Result<never, infer ResultError>
+				? AsyncResult<Value, Err | ResultError>
+				: AsyncResult<Value, Err>
+			: ReturnType extends Result<never, infer ResultError>
+				? AsyncResult<Value, Err | ResultError>
+				: AsyncResult<Value, Err>;
+	}
+
 	/**
 	 * Like {@linkcode AsyncResult.map} it transforms the value of a successful result using the {@link transform} callback.
 	 * In addition, it catches any exceptions that might be thrown inside the {@link transform} callback and encapsulates them
 	 * in a failed result.
 	 *
 	 * @param transform callback function to transform the value of the result. The callback can be async as well.
+	 * @param errorTransform optional callback function to transform the value of the error. The callback can be async as well.
```

Comment:
I don't think we should support an async `errorTransform` (at least for now), and I don't think this implementation supports it either.